File: Week27 Codes/3.Key Sharing/DH.cpp
#include <botan/auto_rng.h>
#include <botan/dh.h>
#include <botan/hex.h>
#include <botan/dl_group.h>
#include <botan/pubkey.h>
#include <iostream>
using namespace Botan;

int main()
{
    AutoSeeded_RNG rng;

    // Generate parameters for DH
    Botan::DL_Group dh_grp("modp/ietf/2048");

    // Generate Alice's keys
    Botan::DH_PrivateKey alice_priv(rng, dh_grp);
    std::vector<uint8_t> alice_pub_key = alice_priv.public_value();

    // Generate Bob's keys
    Botan::DH_PrivateKey bob_priv(rng, DL_Group("modp/ietf/2048"));
    std::vector<uint8_t> bob_pub_key = bob_priv.public_value();

    // Both parties share their public keys
    //  Alice calculates shared key using Bob's public key
    Botan::PK_Key_Agreement alice_agree(alice_priv, rng, "KDF2(SHA-256)");
    Botan::SymmetricKey alice_shared_key = alice_agree.derive_key(32, bob_pub_key);

    // Bob calculates shared key using Alice's public key
    Botan::PK_Key_Agreement bob_agree(bob_priv, rng, "KDF2(SHA-256)");
    Botan::SymmetricKey bob_shared_key = bob_agree.derive_key(32, alice_pub_key);

    // Convert shared keys to hex strings for display
    std::cout << "Alice's Shared Key: " << hex_encode(alice_shared_key) << std::endl;
    std::cout << "Bob's Shared Key: " << hex_encode(bob_shared_key) << std::endl;
    if (alice_shared_key == bob_shared_key)
    {
        std::cout << "Succesfully shared the key\n";
    }
    else
    {
        std::cout << "Key sharing is not succeess\n";
    }

    return 0;
}

File: Week27 Codes/3.Key Sharing/DH.py
import botan3 as botan
import secrets
import os

rng = botan.RandomNumberGenerator()
salt_length = 16  # Adjust the length as needed
salt = os.urandom(salt_length)

# Generate DH parameters for Alice and Bob
alice_priv = botan.PrivateKey.create('DH', 'modp/ietf/2048',rng)
bob_priv = botan.PrivateKey.create('DH', 'modp/ietf/2048',rng)

alice_pub = botan.PKKeyAgreement(alice_priv, 'KDF2(SHA-256)')
bob_pub = botan.PKKeyAgreement(bob_priv, 'KDF2(SHA-256)')

# Alice and Bob compute the shared key using other's public key
alice_shared_key = alice_pub.agree(bob_pub.public_value(),32,salt)
bob_shared_key = bob_pub.agree(alice_pub.public_value(),32,salt)

# Convert shared keys to hex strings for display
alice_shared_key_hex = alice_shared_key.hex()
bob_shared_key_hex = bob_shared_key.hex()

print("Alice's Shared Key:", alice_shared_key_hex)
print("Bob's Shared Key:", bob_shared_key_hex)

# Verify if both shared keys are identical
if (alice_shared_key_hex==bob_shared_key_hex):
    print("Key sharing Succesful")
assert alice_shared_key_hex == bob_shared_key_hex, "Shared keys do not match!"
